src/ledger/LedgerState.cpp

```diff
@@ -1168,66 +1360,82 @@ LedgerStateRoot::getAllOffers()
 std::map<LedgerKey, LedgerEntry>
 LedgerStateRoot::Impl::getAllOffers()
 {
-    auto offers = loadAllOffers();
-    std::map<LedgerKey, LedgerEntry> offersByAccount;
+    std::vector<LedgerEntry> offers;
+    try
+    {
+        offers = loadAllOffers();
+    }
+    catch (...)
+    {
+        throw LedgerStateRootFatalError("LedgerStateRoot could not load from database");
+    }
+
+    std::map<LedgerKey, LedgerEntry> offersByKey;
     for (auto const& offer : offers)
     {
-        offersByAccount[LedgerEntryKey(offer)] = offer;
+        offersByKey.emplace(LedgerEntryKey(offer), offer);
     }
-    return offersByAccount;
+    return offersByKey;
 }
 
-std::shared_ptr<LedgerEntry>
+std::shared_ptr<LedgerEntry const>
 LedgerStateRoot::getBestOffer(Asset const& buying, Asset const& selling,
                               std::set<LedgerKey>&& exclude)
 {
     return mImpl->getBestOffer(buying, selling, std::move(exclude));
 }
 
-std::shared_ptr<LedgerEntry>
+static std::shared_ptr<LedgerEntry const>
+findIncludedOffer(std::list<LedgerEntry>::const_iterator iter,
+                  std::list<LedgerEntry>::const_iterator const& end,
+                  std::set<LedgerKey> const& exclude)
+{
+    for ( ; iter != end; ++iter)
+    {
+        auto key = LedgerEntryKey(*iter);
+        if (exclude.find(key) == exclude.end())
+        {
+            return std::make_shared<LedgerEntry const>(*iter);
+        }
+    }
+    return {};
+}
+
+std::shared_ptr<LedgerEntry const>
 LedgerStateRoot::Impl::getBestOffer(Asset const& buying, Asset const& selling,
                                     std::set<LedgerKey>&& exclude)
 {
-    std::string cacheKey = binToHex(xdr::xdr_to_opaque(buying))
-                         + binToHex(xdr::xdr_to_opaque(selling));
+    auto cacheKey = getBestOffersCacheKey(buying, selling);
     if (!mBestOffersCache->exists(cacheKey))
     {
-        mBestOffersCache->put(cacheKey, {{}, false});
+        putInBestOffersCache(cacheKey, {{}, false});
     }
-    auto& cached = mBestOffersCache->get(cacheKey);
-
+    auto& cached = getFromBestOffersCache(cacheKey);
     auto& offers = cached.bestOffers;
-    for (auto const& offer : offers)
-    {
-        auto key = LedgerEntryKey(offer);
-        if (exclude.find(key) == exclude.end())
-        {
-            return std::make_shared<LedgerEntry>(offer);
-        }
-    }
+
+    auto res = findIncludedOffer(offers.cbegin(), offers.cend(), exclude);
 
     size_t const BATCH_SIZE = 5;
-    while (!cached.allLoaded)
+    while (!res && !cached.allLoaded)
     {
-        auto newOffers =
-            loadBestOffers(buying, selling, BATCH_SIZE, offers.size());
-        if (newOffers.size() < BATCH_SIZE)
+        std::list<LedgerEntry>::const_iterator newOfferIter;
+        try
         {
-            cached.allLoaded = true;
+            newOfferIter =
+                loadBestOffers(offers, buying, selling, BATCH_SIZE, offers.size());
+        }
+        catch (...)
```

Comment:
same thing than earlier: use proper `catch(std::exception)` instead of just `catch(...)`